crates/extensions/c8y_mapper_ext/src/firmware_update.rs

```diff
@@ -0,0 +1,588 @@
+            r#"{"name":"firmware", "version":"1.0"}"#,
+        ))
+        .await
+        .expect("Send failed");
+
+        // Expect auto-registration message
+        assert_received_includes_json(
+            &mut mqtt,
+            [(
+                "te/device/child1//",
+                json!({"@type":"child-device","@id":"test-device:device:child1"}),
+            )],
+        )
+        .await;
+
+        assert_received_contains_str(
+            &mut mqtt,
+            [(
+                "c8y/s/us",
+                "101,test-device:device:child1,child1,thin-edge.io-child",
+            )],
+        )
+        .await;
+
+        // Validate if current installed firmware message is sent
+        assert_received_contains_str(
+            &mut mqtt,
+            [(
+                "c8y/s/us/test-device:device:child1",
+                "115,firmware,1.0,none",
+            )],
+        )
+        .await;
+
+        // Validate if the supported operation file is created
+        assert!(ttd
+            .path()
+            .join("operations/c8y/test-device:device:child1/c8y_Firmware")
+            .exists());
+    }
+
+    #[tokio::test]
+    async fn mapper_converts_smartrest_firmware_req_to_firmware_update_cmd_for_main_device() {
+        let cfg_dir = TempTedgeDir::new();
+        let (mqtt, _http, _fs, _timer, _ul, _dl) = spawn_c8y_mapper_actor(&cfg_dir, true).await;
+        let mut mqtt = mqtt.with_timeout(TEST_TIMEOUT_MS);
+
+        skip_init_messages(&mut mqtt).await;
+
+        // Simulate c8y_Firmware SmartREST request
+        mqtt.send(MqttMessage::new(
+            &C8yTopic::downstream_topic(),
+            "515,test-device,myFirmware,1.0,http://www.my.url",
+        ))
+        .await
+        .expect("Send failed");
+
+        let (topic, received_json) = mqtt
+            .recv()
+            .await
+            .map(|msg| {
+                (
+                    msg.topic,
+                    serde_json::from_str::<serde_json::Value>(msg.payload.as_str().expect("UTF8"))
+                        .expect("JSON"),
+                )
+            })
+            .expect("No message received");
+
+        let (entity, channel) = MqttSchema::default().entity_channel_of(&topic).unwrap();
+        assert_eq!(entity, "device/main//");
+
+        match channel {
+            Channel::Command {
+                operation: OperationType::FirmwareUpdate,
+                ..
+            } => {
+                let expected_json = json!({
+                    "status": "init",
+                    "name": "myFirmware",
+                    "version": "1.0",
+                    "remoteUrl": "http://www.my.url"
+                });
+                assert_json_diff::assert_json_include!(actual: received_json, expected: expected_json);
+            }
+            _ => panic!("Unexpected response on channel: {channel:?}, topic: {topic:?}"),
+        }
+    }
```

Comment:
So much code to just match a topic-prefix and the payload, just because the topic is dynamic with the cmd_id, highlights to the need of a helper function that matches topic prefix instead of the whole topic. That helper would be useful for all such command converter tests for other ops as well.

Reply:
It was kinda my mistake. Since `firmware_update` doesn't produce `tedgeUrl`, so the JSON payload doesn't need to get `cmd_id` to assert. Here I replaced simply with `assert_received_includes_json`.